pkg/lib/dialogs.tsx

```diff
@@ -82,64 +82,69 @@
+ * be used with dialogs shown by `Dialogs.show()`.
  *
- *   function render() {
- *     const Dialogs = this.context;
- *     return <Button onClick={() => this.handleClick}>Open dialog</Button>;
- *   }
- * }
+ * - Dialogs.run(component, {... props})
  *
- * class MyDialog extends React.Component {
- *   static contextType = DialogsContext;
+ * Shows a dialog and asyncronously waits for it to close.  This creates and
+ * shows a MyDialog with the given properties, plus a special "result" property
+ * which has .resolve() and .reject() methods on it.  Calling either of those
+ * will resolve the promise returned by Dialogs.run() accordingly, closing the
+ * dialog in the process.  The created dialog cannot be closed with
+ * Dialogs.close(). See the example:
  *
- *   render() {
- *     <Button onClick={() => Dialogs.close("yes")}>Yes</Button>
- *     <Button onClick={() => Dialogs.close("no")}>No</Button>
- *   }
- * }
+ * const MyDialog = ({ title, result }) => {
+ *     return (
+ *         <Modal title={title}>
+ *             <Button onClick={() => result.resolve("yes")}>Yes</Button>
+ *             <Button onClick={() => result.resolve("no")}>No</Button>
+ *         </Modal>
+ *     );
+ * };
  *
+ * const AsyncDialogExample = () => {
+ *     const Dialogs = useDialogs();
  *
- * - Dialogs.show(component)
+ *     const clicked = async () => {
+ *         try {
+ *             const result = await Dialogs.run(MyDialog, { title: "Example" });
+ *             console.log(result);
+ *         } catch (err) {
+ *         }
+ *     };
  *
- * Calling "Dialogs.show" will render the given component as a direct
- * child of the inner-most enclosing "WithDialogs" component.  The
- * component is of course intended to be a dialog, such as
- * Patternfly's "Modal".  There is only ever one of these; a second
- * call to "show" is considered a bug and "Dialogs.close" should be called first.
- * "Dialogs.show" returns a promise that is settled by either "Dialogs.close" or
- * "Dialogs.reject".
+ *     return <Button onClick={clicked}>Open dialog</Button>;
+ * };
  *
- * - Dialogs.close([args])
+ * - Dialogs.isActive()
  *
- * Calling "Dialogs.close([args])" will close the currently open Dialog and
- * optionally resolve the promise with the provided "args".
+ * Returns `true` if a dialog is currently being shown.
  *
- * - Dialogs.reject(err)
-  *
- * Calling "Dialogs.reject(err)" will close the currently open Dialog
- * and reject the promise with the provided Error.
  */
 
-import React, { useContext, useRef, useState } from "react";
+import React, { useContext, useState, useRef } from "react";
+
+export interface DialogResult<T> {
+    resolve(value: T): void;
+    reject(exc: unknown): void;
+}
 
 export interface Dialogs {
-    show(dialog: React.ReactNode): Promise<unknown>;
-    close(args: unknown): void;
-    reject(err: unknown): void;
+    show(dialog: React.ReactNode): void;
+    close(): void;
```

Comment:
Indeed we never seem to explicitly pass a value to close(), so this API change seems fine.

Reply:
Cockpit Files upload conflict resolution dialog is the one exception.